pymbolic/mapper/__init__.py

```diff
@@ -175,76 +208,135 @@ def rec_fallback(self, expr, *args, **kwargs):
         else:
             return self.map_foreign(expr, *args, **kwargs)
 
-    def map_algebraic_leaf(self, expr, *args, **kwargs):
+    def map_algebraic_leaf(self,
+            expr: primitives.AlgebraicLeaf,
+            *args: P.args, **kwargs: P.kwargs) -> ResultT:
         raise NotImplementedError
 
-    def map_variable(self, expr, *args, **kwargs):
+    def map_variable(self,
+            expr: primitives.Variable, *args: P.args, **kwargs: P.kwargs) -> ResultT:
         return self.map_algebraic_leaf(expr, *args, **kwargs)
 
-    def map_subscript(self, expr, *args, **kwargs):
+    def map_subscript(self,
+            expr: primitives.Subscript, *args: P.args, **kwargs: P.kwargs) -> ResultT:
         return self.map_algebraic_leaf(expr, *args, **kwargs)
 
-    def map_call(self, expr, *args, **kwargs):
+    def map_call(self,
+            expr: primitives.Call, *args: P.args, **kwargs: P.kwargs) -> ResultT:
         return self.map_algebraic_leaf(expr, *args, **kwargs)
 
-    def map_lookup(self, expr, *args, **kwargs):
+    def map_call_with_kwargs(self,
+            expr: primitives.CallWithKwargs,
+            *args: P.args, **kwargs: P.kwargs) -> ResultT:
         return self.map_algebraic_leaf(expr, *args, **kwargs)
 
-    def map_if_positive(self, expr, *args, **kwargs):
+    def map_lookup(self,
+            expr: primitives.Lookup, *args: P.args, **kwargs: P.kwargs) -> ResultT:
         return self.map_algebraic_leaf(expr, *args, **kwargs)
 
-    def map_rational(self, expr, *args, **kwargs):
-        return self.map_quotient(expr, *args, **kwargs)
+    def map_if(self,
+            expr: primitives.If, *args: P.args, **kwargs: P.kwargs) -> ResultT:
+        raise NotImplementedError
 
-    def map_quotient(self, expr, *args, **kwargs):
+    def map_rational(self,
+            expr: Rational, *args: P.args, **kwargs: P.kwargs) -> ResultT:
         raise NotImplementedError
 
-    def map_constant(self, expr, *args, **kwargs):
+    def map_quotient(self,
+            expr: primitives.Quotient, *args: P.args, **kwargs: P.kwargs) -> ResultT:
         raise NotImplementedError
 
-    def map_list(self, expr, *args, **kwargs):
+    def map_floor_div(self,
+            expr: primitives.FloorDiv, *args: P.args, **kwargs: P.kwargs) -> ResultT:
         raise NotImplementedError
 
-    def map_tuple(self, expr, *args, **kwargs):
+    def map_remainder(self,
+            expr: primitives.Remainder, *args: P.args, **kwargs: P.kwargs) -> ResultT:
         raise NotImplementedError
 
-    def map_numpy_array(self, expr, *args, **kwargs):
+    def map_constant(self,
+            expr: object, *args: P.args, **kwargs: P.kwargs) -> ResultT:
```

Comment:
Perhaps it is better to use 

```suggestion
            expr: Any, *args: P.args, **kwargs: P.kwargs) -> ResultT:
```

for derived Mappers?

Reply:
Why do you say so? `Any` will allow anything done to it, `object` will allow very little, so derived mappers are forced to do appropriate `isinstance` checks, which seems good practice to me.

Reply:
But then you get these kinds of errors, needing more type-ignores:

```
pymbolic/interop/ast.py:286: error: Signature of "map_constant" incompatible with supertype "Mapper"  [override]
pymbolic/interop/ast.py:286: note:      Superclass:
pymbolic/interop/ast.py:286: note:          def map_constant(self, expr: object, *args: Any, **kwargs: Any) -> Any
pymbolic/interop/ast.py:286: note:      Subclass:
pymbolic/interop/ast.py:286: note:          def map_constant(self, expr: int | integer[Any] | float | complex | inexact[Any] | bool | bool_) -> expr
```

I'm not sure which way is better.

Reply:
No; the derived mapper needs to change its annotation to `object` and actually handle it.